GUIDE.md

```diff
@@ -0,0 +1,178 @@
+
+`Match` is a newtype that looks like this:
+
+```purescript
+newtype Match a = Match (Route -> V (Free MatchError) (Tuple Route a))
+```
+
+Don't worry too much about the scary type for now.  Essentially, `Match` wraps a function that takes a `Route` (represented as a `List RoutePart`, where `RouteParts` is either the stuff between url slashes or the query parameters) and returns a `V` functor.  `V` stands for 'validation' and you can think of it as an `Either` that can return more than one error upon failure.  So, in essence, something of type `Match Locations` is a function from `Route -> Either Errors (Tuple (Rest-of-RoutePart-List) Locations)`.
+
+The real magic happens in `Match`'s `Apply (i.e., <*>)` implementation.
+
+The `Apply` instance for `Match` provides the plumbing that allows you to compose your routes.  The `<*>`, `*>` and `<*` operators are your friends.  Like any other applicative, `(leftMatch *> rightMatch)` performs the left `Match`, discards a successful result, and returns the value of the right `Match`.  `(leftMatch <* rightMatch)` does the converse.  So, `MPC <$> (homeSlash *> lit "mpc" *> int)` maps the `MPC` constructor function over a `Match` that will parse a slash (and discard it), parse the literal string "mpc" (discarding it too), and return the parsed integer.
+
+Routes are combined using `Match`'s `Alt (<|>)` instance.
+
+```purescript
+routing :: Match Locations
+routing =
+  jupiter <|>
+  mars    <|>
+  mpc     <|>
+  moon    <|>
+  home
+```
+`jupiter (<|>) mars` means "try Jupiter, and if that fails, try Mars".  Like many routing DSLs, this means that more general routes come after more specific ones.  Thus, `home` (i.e., "/") comes last (unless you only ever want to go home).
+
+We've got routes.  Now what do we do with them?
+
+`purescript-routing` provides a couple functions that help you make use of your `Match`s.
+
+```purescript
+matches :: forall e a. Match a -> (Maybe a -> a -> Eff e Unit) -> Eff e Unit
+```
+
+and
+
+```purescript
+matchesAff :: forall e a. Match a -> Aff e (Tuple (Maybe a) a)
+```
+
+`matches` and `matchesAff` provide access to the stream of hash changes.  You provide a function with your `routing` (e.g., your composed `Match Locations`), and `matches` (or `matchesAff`) maps it over the stream.
+
+In our example, we'll run in `Eff`:
+
+```purescript
+main = do
+  matches routing (\old new -> someAction new)
+        --- other stuff ---
+  where
+    someAction :: forall e. Maybe Locations -> Locations -> Eff e Unit
+    someAction = ...
+```
+The previous route (`old`) is a `Maybe Locations` and the `new` is of type `Locations`.  The `someAction` you take is up to you, but you'll probably want to do something with the `new` value.
+
+####Going Deeper
+
+If you're interested in the types and implementation details, here's a sequence of types and functions to get you started.
+
+###### in `Routing.Types`
```

Comment:
I feel like replicating all these types inside the GUIDE.md will make the material go out of date quickly, and place an additional burden on library maintainers. The introduction, however, is quite nice!